### PR TITLE
[develop] Add Community Fire Behavior Model

### LINKED ISSUE #1147: Need an easy way to reference build settings at runtime
## Description
For various reasons, it may be necessary to be able to reference some build settings (e.g. build-time flags, compiler, etc.) when setting up and/or running your SRW experiment.

## Solution
Take advantage of the existing logic in the top-level CMakeLists.txt to write a yaml file with build settings that can then be easily read by the workflow at both setup- and run-time.

For example, with the new UFS fire build (#1139), it is necessary to check that the weather model has been built with fire capabilities before we generate an experiment with fire settings. It could be used similarly in checks for AQM builds; currently this is not done and so error checking is incomplete. This could also be used to deprecate the need to specify a compiler manually in `config.yaml`, avoiding potential errors that come from incorrect user settings. And I'm sure there are many other cases where this capability would be useful.

## Requirements**
Build settings should be in yaml format so that it can be read like any other SRW settings file. Should include the "application" that was built, compiler, and other build-time settings that would be useful to reference at configure/runtime.

## Acceptance Criteria (Definition of Done)
Merging of PR #1139

### PR DESCRIPTION
**Pull request by mkavulich**
## DESCRIPTION OF CHANGES: 

This PR introduces the Community Fire Behavior Module (https://github.com/ufs-community/ufs-weather-model/pull/2220) to the SRW App. In addition, there are a number of general improvements to the UFS SRW code and workflow. All of these changes are described in more detail below.

### Community Fire Behavior Module

Documentation has been added to the Users Guide, but the basics are that the fire build requires an additional build flag

```
./devbuild.sh --platform=[hera|derecho|etc.] --app=ATMF
```

and has an additional `fire:` section in `config.yaml`. There is an example fire configuration available in the new example config `config.fire.yaml`. Many settings are described in detail in `config_defaults.yaml`, but I am working on additional details.

The fire model outputs additional netCDF files with fire-specific information; two example plots of this data can be found here:
ncview emis_smoke_trim
ncview fire_area_trim

Sample data for the case in the example file has been staged on Derecho (/glade/derecho/scratch/kavulich/FIRE/2024/test_data) and Hera (/scratch2/BMC/fv3lam/kavulich/UFS/workdir/UFS_FIRE/test_data). I will reach out to EPIC to see if they would be willing to officially stage this sample data on all Level 1 platforms for community use and potential future tests. This PR does not include a WE2E test because the testing system does not currently have a neat way to handle multiple build types, but in theory the new example case could be used. 

Two new pre-defined domains `SUBCONUS_CO_1km` and `SUBCONUS_CO_3km`, for use with high-resolution fire experiments, are added.

### `build_settings.yaml`
In `develop` there is currently no way to reference build settings at configure and run time. There was a vestigial capability in the `ufs_srweather_app.settings.in` which was added 4+ years ago, but it has never really been used or updated and it is in an ad-hoc format that cannot be easily used. I have leveraged this existing capability to create a new file, `build_settings.yaml` in the build directory that contains useful build information (the application built, the compiler, the git hash, etc.) and can be easily read using uwtools in the existing workflows and run scripts. Specifically this is used in this PR to ensure that the app has been built with the fire behavior capability if fire options are selected, but this should prove very useful in other parts of the workflow going forward; for just a few examples:
 - checking that the right compiler was specified for the experiment settings
 - checking that machine settings are correct compared to the build
 - checking app-specific settings for other builds such as AQM

This new file and the unused (but potentially useful in the future) `ufs_srweather_app_meta.h.in` files are moved to the `sorc` directory to keep the source code more organized and uncluttered.

### Improve capability to use a different set of vertical levels
Currently the [instructions for running SRW with a different number of vertical levels from the default 65](https://ufs-srweather-app.readthedocs.io/en/latest/UsersGuide/CustomizingTheWorkflow/LAMGrids.html#changing-the-number-of-vertical-levels) is quite convoluted, involving the need to manually edit the template namelist file. This PR adds LEVP as a configuration variable to more easily allow for runs with different vertical levels. The users guide instructions have been updated accordingly.

### Flexible config file name
Now rather than being forced to use the name `config.yaml` for your experiment config, you can reference any yaml file by using the `-c` option for workflow generation:

```
./generate_FV3LAM_wflow.py -c config.fire.yaml
```

With these changes, `EXPT_CONFIG_FN` is removed as a configurable option, but this was uselessly self-referential anyway.

### Additional options for `retrieve_data.py`
A big limitation in specifying filenames in the current app is that there is no capability for filenames to change based on the forecast hour vs the cycle hour. Therefore a new set of variables, identical to the originals but with the prefix `f`, can now be used when specifying filenames with respect to the forecast hour rather than the cycle hour.

An example of this new usage can be seen in the newly added `config.fire.yaml` experiment:

```
  EXTRN_MDL_FILES_LBCS:
    - '{fyyyymmdd}.hrrr.t{fhh}z.wrfprsf00.grib2'
```
This instructs `retrieve_data.py` to look for filenames specific to the valid forecast hour for the specific lateral boundary condition file. So for forecast hour zero (valid 2020081318), the expected filename will be `20200813.hrrr.t18z.wrfprsf00.grib2`. But at the 18 hour forecast, the LBC filename will be `20200814.hrrr.t12z.wrfprsf00.grib2`. This was not previously possible without these new variables, and represents a vast improvement in flexibility for users.

### Speedup of symbolic linking
I found that (depending on the platform) the `create_symlink_to_file.sh` function could take as many as several seconds for each link, an absurdly long time for such a simple operation! I removed some superfluous logic from this function and `boolify.sh` (which ensures that booleans are properly compared in bash, a necessary evil), which brought this time down to tenths of seconds. This results in a speedup of a minute or more for the run_fcst task, depending on the platform and the settings.

### Various random improvements
 - Fixed regex statements in WE2E tests to fix `SyntaxWarning: invalid escape sequence`
 - Updated build modulefiles to include `FC` variable, which is needed by the Fire Behavior component.

### Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 
A number of fundamental test runs have been conducted on Derecho, Hera, and Orion, both with and without the ATMF build. Not all are passing, but seems to be unrelated to this development; see the note at the top. This section will be updated once those issues are resolved.

- [x] derecho.intel
  - fundamental tests, both with and without -a=ATMF build
  - new UFS_FIRE test
- [ ] gaea.intel
- [ ] hera.gnu
- [ ] hera.intel
- [ ] hercules.intel
- [ ] jet.intel
- [ ] orion.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
#1136 should be merged first, as it includes a fix for Hera runs with the new UFS weather model hash.

## DOCUMENTATION:
Users guide has been updated with instructions for building and running with these new capabilities; see draft documentation build in the Github checks below. Specifically, [this is the new chapter on the UFS_FIRE capability](https://ufs-srweather-app--1139.org.readthedocs.build/en/1139/UsersGuide/BuildingRunningTesting/FIRE.html).

## ISSUE: 
 - Resolves #1147

## CHECKLIST

- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## CONTRIBUTORS (optional): 
Thanks to the whole Fire Behavior team for bringing me on to this project, and their extensive efforts to get this capability ready for use by the scientific community: @danrosen25 @masih-e @pedro-jm and @mefrediani

### CODE REVIEW

File: devbuild.sh
Comment (christinaholtNOAA) [inline devbuild.sh]: Should this list be pruned to the ones the app supports? 

Do you want to add a blurb about ATMF in the text above the list?

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (christinaholtNOAA) [inline doc/UsersGuide/BuildingRunningTesting/FIRE.rst]:
```suggestion
The biggest difference between the UFS FIRE capability and other modes of the UFS SRW is that a special build flag is required to build the coupled fire behavior code, as described in the instructions below. Aside from that, the need for additional input files, and some fire-specific config settings, configuring and running an experiment is the same as any other use of SRW.
```

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (christinaholtNOAA) [inline doc/UsersGuide/BuildingRunningTesting/FIRE.rst]:
```suggestion
Build the SRW App with Fire Behavior Enabled
```

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (christinaholtNOAA) [inline doc/UsersGuide/BuildingRunningTesting/FIRE.rst]:
```suggestion
Load the appropriate modules for the workflow:
```

This step also loads the Rocoto module.

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (christinaholtNOAA) [inline doc/UsersGuide/BuildingRunningTesting/FIRE.rst:82]: This is no longer necessary, right?

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (christinaholtNOAA) [inline doc/UsersGuide/BuildingRunningTesting/FIRE.rst:163]: This gets a command line argument now, no?

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (christinaholtNOAA) [inline doc/UsersGuide/BuildingRunningTesting/FIRE.rst]:
```suggestion
   These files contain output directly from the fire model (hence the greater frequency), including variables such as the fire perimeter and area, smoke emitted, and fuel percentage burned. 
```

**Review by christinaholtNOAA — CHANGES_REQUESTED** [review]
As promised, you have added a lot of documentation. I only got as far as FIRE.rst in this first round, but will return to it early next week.

These are only minor suggestions. Do with them what you will.

**Comment by mkavulich** [comment]
@MichaelLueken I will need help staging the data for the new UFS_FIRE WE2E test. It is located on Hera at `/scratch2/BMC/fv3lam/kavulich/UFS/workdir/UFS_FIRE/test_data/geo_em.d01.nc` and `/scratch2/BMC/fv3lam/kavulich/UFS/workdir/UFS_FIRE/test_data/HRRR/`; feel free to organize the data in a way that makes sense in the EPIC test data space and I can change the test files to point to it accordingly.

File: doc/UsersGuide/CustomizingTheWorkflow/ConfigWorkflow.rst
Comment (MichaelLueken) [inline doc/UsersGuide/CustomizingTheWorkflow/ConfigWorkflow.rst:1]: Thanks for removing the `EXPT_CONFIG_FN` parameter from the documentation!  Please add the new `LEVP` and `UFS FIRE` configuration parameters to the `ConfigWorkflow.rst` file as well.

**Review by MichaelLueken — CHANGES_REQUESTED** [review]
@mkavulich -

I was able to successfully run the new UFS Fire sample configuration on Derecho.

It would be best to include the new `UFS Fire` configuration parameters, as well as `LEVP`, in the updates to `doc/UsersGuide/CustomizingTheWorkflow/ConfigWorkflow.rst`.

File: parm/namelist.fire
Comment (christinaholtNOAA) [inline parm/namelist.fire]: UW Tools will likely be able to help you remove these hard-coded values with references to a Python datetime object. Rendering the template will allow you to avoid the octal problem you mention in the bash script below.

Here, the changes might look like:

```
&time
  start_year = '{{ cycle.year }}'
  start_month = '{{ cycle.month }}'
  start_day = '{{ cycle.day }}'
  ...
  end_year = '{{ endtime.year }}'
/
```


To render this to stdout, you can use:
```
uw template render --input-file fire.nml --values-format yaml --values-file <(echo '{cycle: !!timestamp "2024-09-10 12:00:00", endtime: !!timestamp "2024-09-15 12:00:00"}')
```

File: parm/ufs.configure
Comment (christinaholtNOAA) [inline parm/ufs.configure:47]: I'm assuming this bit replaces the run sequence stuff deleted above? Nice.

File: scripts/exregional_run_fcst.sh
Comment (christinaholtNOAA) [inline scripts/exregional_run_fcst.sh:783]: Here is where you might change to `uw template render` suggested above, but use the `-o "${FIRE_NML_FN}"` to have it go to the appropriate file instead of stdout.

This would require a change to the error handling below, too.

File: tests/WE2E/WE2E_summary.py
Comment (christinaholtNOAA) [inline tests/WE2E/WE2E_summary.py:15]: Could this not be simplified by using a UW config tool from the API?

File: tests/WE2E/WE2E_summary.py
Comment (christinaholtNOAA) [inline tests/WE2E/WE2E_summary.py:77]: I know this was already here with this name, but somehow putting it in an if-statement like this, I now read "expects dict". Not sure if it matters much.

File: tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml
Comment (christinaholtNOAA) [inline tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml]: I'm not sure you want to carve out the space "special builds" for this. We'd presumably have many different "ufs-weather-model app builds" in the future. Maybe it makes sense to put it under a subdir `test_configs/ufs_fire/`, then we may have several test configs for the UFS Fire build, and others under similarly named `test_configs/mpas_dycore`, or `test_configs/aqm`, etc.

File: tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml
Comment (christinaholtNOAA) [inline tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml]: I assume this would eventually live under the EPIC role account space on the Tier I platforms?

File: ush/config_defaults.yaml
Comment (christinaholtNOAA) [inline ush/config_defaults.yaml]: Just to keep it clearer, would you mind changing it so that the `fire.FIRE_NUM_TASKS` entry ends up in the same position in both of these equations? I personally prefer seeing it show up at the end, but consistency is preferable either way.

File: ush/create_ufs_configure_file.py
Comment (christinaholtNOAA) [inline ush/create_ufs_configure_file.py]:
```suggestion
    # Call the uwtools "render" function to fill in jinja expressions contained in the UFS Configure file template
    # with the values from the settings variable to create ufs.configure file for this experiment
```

Would you mind using the term `jinja expressions` here? It would make it more consistent with uwtools documentation and may help users who are cross-referencing docs. In general, we've been trying to refer to the `{{ }}` entries as Jinja expressions, and then the file that contains them would be a template file.

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [inline ush/generate_FV3LAM_wflow.py:856]: 🎉 🥂 👍 🥇

File: ush/predef_grid_params.yaml
Comment (christinaholtNOAA) [inline ush/predef_grid_params.yaml:339]: I'm assuming this shows up in the test suite now?

File: ush/setup.py
Comment (christinaholtNOAA) [inline ush/setup.py:1524]: Nifty check!

**Review by christinaholtNOAA — COMMENTED** [review]
This is a really nice addition of a new feature and some great upgrades for SRW.

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (mkavulich) [reply doc/UsersGuide/BuildingRunningTesting/FIRE.rst:82]: I'd like to keep this as-is; it's a good idea to not work on top of the example file since that is versioned in the repository.

File: tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml
Comment (mkavulich) [reply tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml]: Yes, I believe EPIC is working on staging the data in the appropriate locations; I will update the test file as necessary when that is complete.

File: tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml
Comment (mkavulich) [reply tests/WE2E/test_configs/special_builds/config.UFS_FIRE.yaml]: Your reasoning makes sense, I'll revert this change.

File: parm/ufs.configure
Comment (mkavulich) [reply parm/ufs.configure:47]: That's correct; the run sequence for each component is now specified in `create_ufs_configure_file.py`

File: ush/predef_grid_params.yaml
Comment (mkavulich) [reply ush/predef_grid_params.yaml:339]: Yes, the test is `grid_RRFS_AK_3km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR`

File: devbuild.sh
Comment (mkavulich) [reply devbuild.sh]: Done

**Comment by mkavulich** [comment]
@MichaelLueken I have added documentation of the config.yaml options for the new Fire capabilities and LEVP. In the future it might be nice to consolidate this information, as it is currently duplicated across the Users guide and `config_defaults.yaml`.

File: parm/namelist.fire
Comment (mkavulich) [reply parm/namelist.fire]: Per our offline discussion, I wasn't able to get this working due to a potential `uw template` issue. I did remove the time values from the base namelist however.

File: tests/WE2E/WE2E_summary.py
Comment (mkavulich) [reply tests/WE2E/WE2E_summary.py:15]: Most likely, but incorporating uw tools into the WE2E tests is beyond the scope of this PR

File: doc/UsersGuide/BuildingRunningTesting/FIRE.rst
Comment (mkavulich) [reply doc/UsersGuide/BuildingRunningTesting/FIRE.rst:163]: It's not mandatory so I don't want to clutter up this specific doc too much, but I will update the docs elsewhere for this new option.

**Comment by mkavulich** [comment]
@christinaholtNOAA @MichaelLueken I believe I have addressed all your comments. Note that I did update how the fire ignition variables are handled in `setup.py` and `generate_FV3LAM_wflow.py` slightly, and add a new WE2E test that covers multiple ignitions. You can see those changes in [this diff](https://github.com/ufs-community/ufs-srweather-app/pull/1139/files/85ca43eb8ee2939198c3281ec30bec774c73dafa..d4043bd2e36748d37ee9883e990f3271d2241f61).

Let me know if you have any more comment/concerns.

File: ush/setup.py
Comment (MichaelLueken) [inline ush/setup.py:1560]: The new WE2E tests and the `config.fire.yaml` sample configuration set `FIRE_ATM_FEEDBACK` to `1.0`.  This is causing the tests to fail since it is larger than `0.0`.  Should the test configurations be updated to use `FIRE_ATM_FEEDBACK: 0.0`, rather than `1.0`?

**Review by MichaelLueken — COMMENTED** [review]
@mkavulich - While attempting to run the two new WE2E tests, they failed to submit because `FIRE_ATM_FEEDBACK > 0.0` for these new configurations.  Should the test configuration use `FIRE_ATM_FEEDBACK: 0.0` instead of `FIRE_ATM_FEEDBACK: 1.0`?

File: ush/setup.py
Comment (mkavulich) [reply ush/setup.py:1560]: Sorry about that, I updated the settings in the WE2E tests but failed to push it here. This is an issue that we found after merging the capability into the weather model where the two-way coupling was not actually working (results were identical to 1-way coupling); this just ensures that users don't try to use two-way coupling while it's not working in the initial version.

Comment (MichaelLueken) [reply ush/setup.py:1560]: That makes sense to me!  This has been resolved as of [e67173f](https://github.com/ufs-community/ufs-srweather-app/pull/1139/commits/e67173fb206c3d38bf78318b8eebfa5c4d0be035).

**Review by MichaelLueken — APPROVED** [review]
Thanks, @mkavulich!  The tests are successfully generated and run.  Also, thank you very much for updating the documentation in `ConfigWorkflow.rst`.  We'll need to see if there is some way to pull information directly from `config_defaults.yaml` into the `ConfigWorkflow.rst` file so that we don't need to repeat modifications to both.

**Comment by mkavulich** [comment]
@MichaelLueken Thanks for the review. Let me know when the data is staged for the WE2E tests; I'll likely need to make some modifications to those test files to point to the correct data.

**Comment by MichaelLueken** [comment]
@mkavulich -

The UFS Fire data has been staged on Tier-1 platforms:
- Hera - /scratch1/NCEPDEV/nems/role.epic/UFS_SRW_data/develop/UFS_FIRE
- Derecho - /glade/work/epicufsrt/contrib/UFS_SRW_data/develop/UFS_FIRE
- Gaea - /gpfs/f5/epic/world-shared/UFS_SRW_data/develop/UFS_FIRE
- Hercules/Orion - /work/noaa/epic/role-epic/contrib/UFS_SRW_data/develop/UFS_FIRE
- Jet - /mnt/lfs5/HFIP/hfv3gfs/role.epic/UFS_SRW_data/develop/UFS_FIRE

And many thanks to @RatkoVasic-NOAA for staging the data across the platforms!

**Comment by mkavulich** [comment]
@MichaelLueken I pushed the final changes for the new fire WE2E tests, I believe this PR is ready for merging presuming all tests pass and @christinaholtNOAA approves.

**Review by christinaholtNOAA — APPROVED** [review]
LGTM.

**Comment by MichaelLueken** [comment]
@mkavulich -

The Jenkins tests are all failing in the `Functional WorkflowTaskTests` phase of the pipeline.  The Jenkins tests don't have an `exec` directory, but an `install_${compiler}/exec` directory.  The `ush/setup.py` script needs to have the `build_config_fp` variable updated to point to either `exec` (standard build) or `install_${compiler}/exec` (Jenkins build), otherwise the Jenkins tests will be unable to run.

**Comment by MichaelLueken** [comment]
@mkavulich -

On Hera, you can see the directory structure for GNU - `/scratch2/NAGAPE/epic/role.epic/jenkins/workspace/s-srweather-app_pipeline_PR-1139/hera` - and Intel - `/scratch2/NAGAPE/epic/role.epic/jenkins/workspace/s-srweather-app_pipeline_PR-1139__2/hera`.  We'll need to ensure that the new `build_settings.yaml` file can be located, whether the file is in `exec` or in `some_user_defined_directory/exec`.

**Comment by mkavulich** [comment]
@MichaelLueken Thanks for the update. I pushed a fix that should work, can you try again?

**Comment by MichaelLueken** [comment]
Thanks, @mkavulich!  The Jenkins tests have been resubmitted.

**Comment by MichaelLueken** [comment]
@mkavulich -

The Jenkins tests are now running without issue, but there is a conflict in `doc/conf.py` where Gillian removed the `Options for autoyaml extension` section that you had modified in this PR.

**Comment by MichaelLueken** [comment]
@mkavulich -

All Jenkins WE2E tests have successfully passed!  I will be able to merge this PR once the conflict in `doc/conf.py` has been resolved.

**Pull request merged**
